### PR TITLE
Cache results of `polars.datatypes.is_polars_dtype()`

### PR DESCRIPTION
**Pull request by jakob-keller**
Optimize `polars._construction._unpack_schema()` by caching results of `polars.datatypes.is_polars_dtype()`. This avoids expense and repetitive calls to is `instance()`.

See also #11043, which is replaced by this PR.

### CODE REVIEW

File: py-polars/polars/utils/_construction.py [hunk 1]
```diff
@@ -664,6 +665,11 @@ def _post_apply_columns(
     return pydf
 
 
+@functools.lru_cache(maxsize=128)
```
Comment (svaningelgem) [inline py-polars/polars/utils/_construction.py:668]: Do we really need to add another method here?

Why can't you put the cache on the `is_polar_dtype` method? Is there any reason for having 2 different methods doing exactly the same?

That would make this PR a 2-line change (import + decorator on the original method), solving the issue as well?

Comment (jakob-keller) [reply py-polars/polars/utils/_construction.py:668]: > Do we really need to add another method here?

That is a valid question. My motivation is that this cache speeds up calls to `pl.from_records()` by ~10% in my use case.
 
> Why can't you put the cache on the `is_polar_dtype` method? Is there any reason for having 2 different methods doing exactly the same?

`is_polars_dtype()` may be called with non-hashable arguments in other code paths, breaking the decorated function. I experimented with a holistic solution in #11054, but the try-except overhead appears to outweigh any performance gains.

Comment (svaningelgem) [reply py-polars/polars/utils/_construction.py:668]: > `is_polars_dtype()` may be called with non-hashable arguments in other code paths, breaking the decorated function. I experimented with a holistic solution in #11054, but the try-except overhead appears to outweigh any performance gains.

Thanks for the response. I didn't consider these indeed...

Comment (ritchie46) [reply py-polars/polars/utils/_construction.py:668]: @jakob-keller can you add a comment on the rationale of having 2 methods.

Comment (jakob-keller) [reply py-polars/polars/utils/_construction.py:668]: > @jakob-keller can you add a comment on the rationale of having 2 methods.

Will do

Comment (jakob-keller) [reply py-polars/polars/utils/_construction.py:668]: Have a look at #11068 for a more comprehensive approach.

**Pull request closed without merging**